### ISSUE TITLE
Delivery tag should be returned as a Math::UInt64 object

### ISSUE THREAD

**Issue by autarch** [markwellis/net-amqp-rabbitmq#30]
If you agree once I get some more round tuits I can probably patch this. I have experience integrating with Math::Int64 from an XS API.

**Comment by manchicken** [comment]
You mean `Math::Int64`, not `Math::UInt64`, yes? (Domains: https://www.rabbitmq.com/amqp-0-9-1-reference.html)

I think that makes sense, especially since it'd keep things in the XS where it belongs.

**Comment by autarch** [comment]
In the librabbitmq code i see them using the `uint64_t` type for the delivery tag.

**Comment by autarch** [comment]
But you're right, the AMQP docs seems to suggest it could be signed (which doesn't make much sense)

**Comment by sbasmanov** [comment]
After changing delivery_tag to numeric type (instead of string), ack() works as expected.
But there is another problem: if I ack() with incorrect tag (let's say 2 instead of expected 1), entire script hangs on next recv() (but ack() succeed without eny error) . With tcpdump I see server returns error: 'PRECONDITION_FAILED - unknown delivery tag 2', but seems it never returned as error by ack().

**Comment by autarch** [comment]
@sbasmanov - without seeing the changes you made it's impossible to know what's happening.

**Comment by manchicken** [comment]
This is a known issue, it's because you're seeing the server respond to an
ack with an unknown delivery tag (for some reason it's considered unknown),
but that response is asynchronous so it shows up on the wire and confuses
the C library.

I submitted this bug for the same problem when it happens on basic.publish:
https://github.com/alanxz/rabbitmq-c/issues/232

On Thu, Jan 15, 2015 at 2:39 PM, Dave Rolsky <notifications@github.com>
wrote:

> @sbasmanov <https://github.com/sbasmanov> - without seeing the changes
> you made it's impossible to know what's happening.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/markwellis/net-amqp-rabbitmq/issues/30#issuecomment-70147717>
> .
>



-- 
~ Michael D. Stemle, Jr.

**Comment by sbasmanov** [comment]
Ok, I see why this happen. So, how we can workaround this? My situation is simple - I have recv() in loop, I get message, process it, issue ack() and then again recv(). So, after I sent wrong delivery_tag in ack(), I see consumer disappear from queue(on management page), method is_connected() returns true and then recv() hangs. Yes, I know that I must issue ack() with proper delivery_tag only, but in case of mistake, I would like to have my script working, rather than hanging on recv(). Any ideas?

**Comment by abraxxa** [comment]
Sounds like if you fix `is_connected` and the error handling is already working as it should, the bug should be fixed.

**Comment by sbasmanov** [comment]
Well, according to alanxz reply, this is more problem with library (RPC-like) design. What I think we can try is to do what he suggests in is_connected() method. Quote:
"A workaround for your case is to occasionally try and read from the channel (amqp_simple_wait_frame_on_channel()) with a short timeout looking for those channel.close messages (if you're using publisher confirms, this is also where you might wait for basic.ack's from the broker)."
So, if is_connected() will be called every time before doing anything, I think this will keep us safe from such "late" responses.

**Comment by manchicken** [comment]
Yeah, @sbasmanov is correct. This isn't anything to do with having lost the connection, the problem is to do with the server having sent something on a connection that is still live which indicates an error, but we weren't expecting that to be on the wire and the client had moved on with another task.

**Comment by manchicken** [comment]
@sbasmanov, if you're having a hang in `recv()`, please submit your sample for how to replicate the problem in a separate issue, this thread should really be used for what @autarch was talking about.

**Comment by sbasmanov** [comment]
Ok, will open separate issue. As to delivery tag, this is what I changed to make it work (just quick fix)
void
net_amqp_rabbitmq_ack(conn, channel, delivery_tag, multiple = 0)
  Net::AMQP::RabbitMQ conn
  int channel
  U32 delivery_tag
  int multiple
  PREINIT:
  CODE:
    die_on_error(aTHX_ amqp_basic_ack(conn, channel, delivery_tag, multiple), conn,
                 "ack");

And in internal_recv():
hv_store(RETVAL, "delivery_tag", strlen("delivery_tag"), newSViv(d->delivery_tag), 0);
instead of
hv_store(RETVAL, "delivery_tag", strlen("delivery_tag"), newSVpvn((const char *)&d->delivery_tag, sizeof(d->delivery_tag)), 0);

Now I have proper delivery_tags.

**Comment by manchicken** [comment]
What @autarch suggested would fix the issue that you seem to have identified here, I'm pretty sure.

**Comment by autarch** [comment]
The delivery tag is a 64-bit value. The librabbitmq library uses an unsigned value, though the AMQP protocol docs suggest it's signed. Either way, a 32-bit Perl is unable to handle 64-bit values properly in an  NV (it can only handle up to 2**53 or so, IIRC).

It might be better to just use the native IV value if we're on a 64-bit Perl though. The things that Math::Int64 corrects for a 64-bit Perl are all about math operations, and you shouldn't be doing math on the delivery tag.

**Comment by sbasmanov** [comment]
@autarch, I'm not familiar with all those perl internals, so I just made quick fix to continue development of my application. I spent a lot of time trying to figure out why delivery_tag is empty and how do I ack() in this case :)

**Comment by autarch** [comment]
@sbasmanov - your code will work fine first c. 4 billion messages or so, then something will blow up.

**Comment by sbasmanov** [comment]
@autarch, so in this case I will have to check value and re-open channel when it's close to overflow :)
But I hope there will be patch for 64bit value.

**Comment by noshankus** [comment]
Hi Guys, I'm running into this issue as well. When I do a tcpdump, I can see the delivery_tag, but if I explode out the hash returned from recv(), the delivery_tag is empty.

Making the assumption that the delivery_tag has a value of 1 for the purposes of a test case, I try to send the ack with a numeric value of 1, which throws the error:

    bad tag at ./amqpConsumer.pl line 237, <PIDFILE> line 1.

Any ideas when a patch could be made for this?

**Comment by sbasmanov** [comment]
@noshankus, in RabitMQ.xs I changed:
//      hv_store(RETVAL, "delivery_tag", strlen("delivery_tag"), newSVpvn((const char *)&d->delivery_tag, sizeof(d->delivery_tag)), 0);
      hv_store(RETVAL, "delivery_tag", strlen("delivery_tag"), newSViv(d->delivery_tag), 0);

and

void
net_amqp_rabbitmq_ack(conn, channel, delivery_tag, multiple = 0)
  Net::AMQP::RabbitMQ conn
  int channel
  U32 delivery_tag
  int multiple
  PREINIT:
  CODE:
    die_on_error(aTHX_ amqp_basic_ack(conn, channel, delivery_tag, multiple), conn,
                 "ack");


This is just temporary workaround. Search for string 'delivery_tag' in RabbitMQ.xs and You will see where changes have to be done.

**Comment by noshankus** [comment]
@sbasmanov , great! :+1: I followed your patch and it does indeed work as expected! Thanks for the quick help :)

This to me is a huge bug. I just hope the guys patching can squeeze this into the next release :8ball:

**Comment by manchicken** [comment]
That may not be 64 bit safe, so do be careful.

**Comment by sbasmanov** [comment]
@manchicken, ok, can we write like this? I didn't test, but it compiles fine.
Should be added #include <stdlib.h>
Anything else is left untouched.

void
net_amqp_rabbitmq_ack(conn, channel, delivery_tag, multiple = 0)
  Net::AMQP::RabbitMQ conn
  int channel
  SV *delivery_tag
  int multiple
  PREINIT:
    STRLEN len;
    uint64_t tag;
    unsigned char *str, *end;
  CODE:
    str = SvPV(delivery_tag, len);
//    if(len != sizeof(tag)) Perl_croak(aTHX_ "bad tag");
//    memcpy(&tag, l, sizeof(tag));
    tag = strtoull(str, &end, 10);
    if (tag == 0 && end == str) {
    /* str was not a number */
Perl_croak(aTHX_ "bad tag: not a number");
    } else if (tag == ULLONG_MAX) {
        /* the value of str does not fit in unsigned long long */
Perl_croak(aTHX_ "bad tag: value too big");
    } else if (*end) {
        /* str began with a number but has junk left over at the end */
Perl_croak(aTHX_ "bad tag: not a number");
    }....
    die_on_error(aTHX_ amqp_basic_ack(conn, channel, tag, multiple), conn,
                 "ack");

**Comment by autarch** [comment]
Perl cannot handle 64-bit integers properly unless it was compiled as a 64-bit Perl. This is why I said this should use Math::Int64.

**Comment by sbasmanov** [comment]
@autarch, well, in this code, delivery_tag is being passed as string, and converted to uint64_t right before calling amq_basic_ack(). So, there is nothing to do for Perl with 64bit values. All done by C.

**Comment by noshankus** [comment]
@sbasmanov @manchicken 
Also note... while the above works, it only works for "consume" and not basic "get".

**Comment by sbasmanov** [comment]
Hmmm.. Just noticed that in internal_recv() d->delivery_tag have type uint64_t. It must be converted to string here, before storing it in RETVAL.

**Comment by sbasmanov** [comment]
So, here is how I see it. What do You think? I just made quick test and it works as expected.
```
--- RabbitMQ.xs.orig	2015-02-04 21:01:40.128413346 +0000
+++ RabbitMQ.xs	2015-02-04 21:00:44.752568019 +0000
@@ -6,6 +6,9 @@
 #include "amqp_tcp_socket.h"
 #include "amqp_private.h"
 
+#include <stdlib.h>
+#include <inttypes.h>
+
 #define __REAL__DEBUG__(X)  X
 #define __DEBUG__(X) /* NOOP */
 
@@ -195,6 +198,7 @@
   SV *hvalue = (SV*)&PL_sv_undef;
   HV *headers = (HV*)&PL_sv_undef;
   amqp_table_entry_t *header_entry = (amqp_table_entry_t*)NULL;
+  char buf[32];
 
   result = 0;
   while (1) {
@@ -215,7 +219,9 @@
       if (frame.frame_type != AMQP_FRAME_METHOD) continue;
       if (frame.payload.method.id != AMQP_BASIC_DELIVER_METHOD) continue;
       d = (amqp_basic_deliver_t *) frame.payload.method.decoded;
-      hv_store(RETVAL, "delivery_tag", strlen("delivery_tag"), newSVpvn((const char *)&d->delivery_tag, sizeof(d->delivery_tag)), 0);
+      memset(buf,0,32);
+      snprintf(buf, 32, "%llu", (unsigned long long)d->delivery_tag);
+      hv_store(RETVAL, "delivery_tag", strlen("delivery_tag"), newSVpvn(buf, strlen(buf)), 0);
       hv_store(RETVAL, "redelivered", strlen("redelivered"), newSViv(d->redelivered), 0);
       hv_store(RETVAL, "exchange", strlen("exchange"), newSVpvn(d->exchange.bytes, d->exchange.len), 0);
       hv_store(RETVAL, "consumer_tag", strlen("consumer_tag"), newSVpvn(d->consumer_tag.bytes, d->consumer_tag.len), 0);
@@ -1139,11 +1145,17 @@
   PREINIT:
     STRLEN len;
     uint64_t tag;
-    unsigned char *l;
+    char *str, *end;
   CODE:
-    l = SvPV(delivery_tag, len);
-    if(len != sizeof(tag)) Perl_croak(aTHX_ "bad tag");
-    memcpy(&tag, l, sizeof(tag));
+    str = SvPV(delivery_tag, len);
+    tag = strtoull(str, &end, 10);
+    if (tag == 0 && end == str) {
+	Perl_croak(aTHX_ "bad tag: not a number");
+    } else if (tag == ULLONG_MAX) {
+	Perl_croak(aTHX_ "bad tag: value too big");
+    } else if (*end) {
+	Perl_croak(aTHX_ "bad tag: bad number");
+    }
     die_on_error(aTHX_ amqp_basic_ack(conn, channel, tag, multiple), conn,
                  "ack");
 
@@ -1155,13 +1167,19 @@
  SV *delivery_tag
  int requeue
  PREINIT:
-   STRLEN len;
-   uint64_t tag;
-   unsigned char *l;
+    STRLEN len;
+    uint64_t tag;
+    char *str, *end;
  CODE:
-   l = SvPV(delivery_tag, len);
-   if(len != sizeof(tag)) Perl_croak(aTHX_ "bad tag");
-   memcpy(&tag, l, sizeof(tag));
+    str = SvPV(delivery_tag, len);
+    tag = strtoull(str, &end, 10);
+    if (tag == 0 && end == str) {
+	Perl_croak(aTHX_ "bad tag: not a number");
+    } else if (tag == ULLONG_MAX) {
+	Perl_croak(aTHX_ "bad tag: value too big");
+    } else if (*end) {
+	Perl_croak(aTHX_ "bad tag: bad number");
+    }
    die_on_error(aTHX_ amqp_basic_reject(conn, channel, tag, requeue), conn,
                 "reject");
```

**Comment by manchicken** [comment]
No way, we cannot reject because the value is a 64-bit number. The spec calls for an int64, and Dave's solution would solve this problem the correct way. I think we should do this the way Dave suggested. @autarch, are you working on implementing the change you suggested?

**Comment by sbasmanov** [comment]
@manchicken, what You mean 'reject because the value is a 64-bit number' ? I'm not rejecting anything.

**Comment by autarch** [comment]
@manchicken - I haven't started on it, but it's really pretty easy to do. I can try to find some round tuits next week at $work, I think.

**Comment by sbasmanov** [comment]
Oh, btw. Even if spec says it should int64, underlying librabbitmq uses uint64_t. And if we cast it to int64, we'll loose some data. So, I think we should look at C lib we're using, rather than specs.

**Comment by manchicken** [comment]
@sbasmanov, sorry, I read `ULONG_MAX`. I still don't think that's the right solution, I think @autarch's is the right one.

@sbasmanov, that's a fine point about the uint64_t, and I didn't see that when @autarch first proposed using `Math::UInt64`. We should stick to the C library since that's what we're actually implementing against.

**Comment by autarch** [comment]
Yes, that's why I said Math::**U**Int64 and not Math::Int64.

**Comment by sbasmanov** [comment]
@manchicken, because I have no idea how to make this using Math::UInt64, I did this simple patch. At least it works for now, while we're waiting for @autarch solution. I agree that using Math::UInt64 would be better, but that's out of scope of my knowledge :)

**Comment by noshankus** [comment]
@autarch @manchicken @sbasmanov As a very basic tester for you guys (I'm not great with C at all, so my contributions on that front can be minimal at best), It seems that the best option has been identified and better still, it makes sense. I can confirm however, that @sbasmanov patch above does indeed work. Albeit a temporary workaround.

However, I am still confused slightly as to the differences between the "consume" method which now works due to the above patch, and the fact that the "get" method still does not. Is the get method defined somewhere else? Is this something that can be fixed in parallel to the "consume" method? Or is that inside the base library?

**Comment by sbasmanov** [comment]
@noshankus, with consume You have to use recv(). If You want to use get(), don't call consume. I'm not sure if this is how it should work, but that's what I found when doing tests.

**Comment by noshankus** [comment]
@sbasmanov Yes, I do this already:

When I "consume":

    $mq->consume($amqpChannel, $queueName, { no_ack => false });
    while($amqpMessage = $mq->recv($queueTimeout)){
        $mq->ack($amqpChannel, $amqpMessage->{delivery_tag}, false);
    }

And when I "get":

    $amqpMessage = $mq->get($amqpChannel, $queueName, { no_ack => 0 });
    print Dumper($amqpMessage);

For the "get", the delivery_tag still is not working.

**Comment by sbasmanov** [comment]
@noshankus, I see now.
In net_amqp_rabbitmq_get:
If You used first solution I suggested with U32 delivery_tag
comment out: hv_store(hv, "delivery_tag", strlen("delivery_tag"), newSVpvn((const char *)&ok->delivery_tag, sizeof(ok->delivery_tag)), 0);
and put hv_store(RETVAL, "delivery_tag", strlen("delivery_tag"), newSViv(d->delivery_tag), 0); instead.

**Comment by sbasmanov** [comment]
Here is final diff, until @autarch will make it with Math::UInt64.

```
--- RabbitMQ.xs.orig	2015-02-05 10:53:00.692180611 +0000
+++ RabbitMQ.xs	2015-02-05 10:55:49.783721757 +0000
@@ -6,6 +6,8 @@
 #include "amqp_tcp_socket.h"
 #include "amqp_private.h"
 
+#include <stdlib.h>
+
 #define __REAL__DEBUG__(X)  X
 #define __DEBUG__(X) /* NOOP */
 
@@ -195,6 +197,7 @@
   SV *hvalue = (SV*)&PL_sv_undef;
   HV *headers = (HV*)&PL_sv_undef;
   amqp_table_entry_t *header_entry = (amqp_table_entry_t*)NULL;
+  char buf[32];
 
   result = 0;
   while (1) {
@@ -215,7 +218,9 @@
       if (frame.frame_type != AMQP_FRAME_METHOD) continue;
       if (frame.payload.method.id != AMQP_BASIC_DELIVER_METHOD) continue;
       d = (amqp_basic_deliver_t *) frame.payload.method.decoded;
-      hv_store(RETVAL, "delivery_tag", strlen("delivery_tag"), newSVpvn((const char *)&d->delivery_tag, sizeof(d->delivery_tag)), 0);
+      memset(buf, 0, sizeof(buf));
+      snprintf(buf, sizeof(buf), "%llu", (unsigned long long)d->delivery_tag);
+      hv_store(RETVAL, "delivery_tag", strlen("delivery_tag"), newSVpvn(buf, strlen(buf)), 0);
       hv_store(RETVAL, "redelivered", strlen("redelivered"), newSViv(d->redelivered), 0);
       hv_store(RETVAL, "exchange", strlen("exchange"), newSVpvn(d->exchange.bytes, d->exchange.len), 0);
       hv_store(RETVAL, "consumer_tag", strlen("consumer_tag"), newSVpvn(d->consumer_tag.bytes, d->consumer_tag.len), 0);
@@ -1139,11 +1144,17 @@
   PREINIT:
     STRLEN len;
     uint64_t tag;
-    unsigned char *l;
+    char *str, *end;
   CODE:
-    l = SvPV(delivery_tag, len);
-    if(len != sizeof(tag)) Perl_croak(aTHX_ "bad tag");
-    memcpy(&tag, l, sizeof(tag));
+    str = SvPV(delivery_tag, len);
+    tag = strtoull(str, &end, 10);
+    if (tag == 0 && end == str) {
+	Perl_croak(aTHX_ "bad tag: not a number");
+    } else if (tag == ULLONG_MAX) {
+	Perl_croak(aTHX_ "bad tag: value too big");
+    } else if (*end) {
+	Perl_croak(aTHX_ "bad tag: bad number");
+    }
     die_on_error(aTHX_ amqp_basic_ack(conn, channel, tag, multiple), conn,
                  "ack");
 
@@ -1157,11 +1168,17 @@
  PREINIT:
     STRLEN len;
     uint64_t tag;
-    unsigned char *l;
+    char *str, *end;
  CODE:
-   l = SvPV(delivery_tag, len);
-   if(len != sizeof(tag)) Perl_croak(aTHX_ "bad tag");
-   memcpy(&tag, l, sizeof(tag));
+    str = SvPV(delivery_tag, len);
+    tag = strtoull(str, &end, 10);
+    if (tag == 0 && end == str) {
+	Perl_croak(aTHX_ "bad tag: not a number");
+    } else if (tag == ULLONG_MAX) {
+	Perl_croak(aTHX_ "bad tag: value too big");
+    } else if (*end) {
+	Perl_croak(aTHX_ "bad tag: bad number");
+    }
    die_on_error(aTHX_ amqp_basic_reject(conn, channel, tag, requeue), conn,
                 "reject");
 
@@ -1283,6 +1300,7 @@
     amqp_rpc_reply_t r;
     int no_ack = 1;
   CODE:
+    char buf[32];
     if(options)
       int_from_hv(options, no_ack);
     amqp_maybe_release_buffers(conn);
@@ -1292,7 +1310,9 @@
       HV *hv;
       amqp_basic_get_ok_t *ok = (amqp_basic_get_ok_t *)r.reply.decoded;
       hv = newHV();
-      hv_store(hv, "delivery_tag", strlen("delivery_tag"), newSVpvn((const char *)&ok->delivery_tag, sizeof(ok->delivery_tag)), 0);
+      memset(buf, 0, sizeof(buf));
+      snprintf(buf, sizeof(buf), "%llu", (unsigned long long)ok->delivery_tag);
+      hv_store(hv, "delivery_tag", strlen("delivery_tag"), newSVpvn(buf, strlen(buf)), 0);
       hv_store(hv, "redelivered", strlen("redelivered"), newSViv(ok->redelivered), 0);
       hv_store(hv, "exchange", strlen("exchange"), newSVpvn(ok->exchange.bytes, ok->exchange.len), 0);
       hv_store(hv, "routing_key", strlen("routing_key"), newSVpvn(ok->routing_key.bytes, ok->routing_key.len), 0);


```

**Comment by noshankus** [comment]
@sbasmanov Excellent :)

I can confirm that the patch above works for both "consume" and "get".
Your line numbers seem to be slightly off however. You using the latest Net-AMQP-RabbitMQ-0.010000 build?

**Comment by sbasmanov** [comment]
@noshankus, no, I have Net-AMQP-RabbitMQ-0.008000 with updated librabbitmq. So, this is why line numbers can be messed up. I'll update this diff for latest version later today.

**Comment by markwellis** [comment]
if the c library is using a uint64, but the spec says it should be a regular int64, is that not a bug in the library?

**Comment by manchicken** [comment]
Yeah, it is, I think.

**Comment by sbasmanov** [comment]
@markwellis, I'm sure somebody didn't pay attention when documenting. Negative delivery tags have no sense. If You look at server source (search for next_tag and DeliveryTag), You'll see that tag always start at 1 (in init) and incremented on each message. Do You really think somebody been writing this code keeping negative numbers in mind? So, I think author of C library had same question and we see it resolved to unsigned long long. That's makes sense.

**Comment by markwellis** [comment]
i agree it doesn't make sense, and that we should follow the c library, and
that the spec is wrong.

but that said, the spec is the spec, so if the c library doesn't follow it
when it should, that's a bug. where it's a bug is the real question

**Comment by sbasmanov** [comment]
@markwellis, I think best way is to look at source. I know nothing about erlang, but I see that delivery_tag is used only in rabbit_channel.erl. But I can't find what type it is. Would be great if someone with erlang knowledge would look at source and find out, what type it is. Then we will know where to submit bug report - rabbitmq community due to bug in spec, or to C lib author.

**Comment by markwellis** [comment]
@sbasmanov, good plan

**Comment by sbasmanov** [comment]
Sent email to Simon MacMullen. Let's see what he say.

**Comment by simonmacmullen** [comment]
Hello! I'm not actually one of the authors of the spec but hopefully I can shed some light on the subject. Erlang integers are always signed and auto-promote to bignums. So that doesn't tell you much; we need to look at the spec.

The AMQP 0-9-1 spec section 4.2.5.1 says "Integers and string lengths are always unsigned and held in network byte order." Note that in this context the spec is referring to integers in methods (see the section 4.2.5 header). So that's what delivery_tag is, an unsigned 64 bit integer.

Confusion might arise though because of our errata: http://www.rabbitmq.com/amqp-0-9-1-errata.html#section_3 which talks about the differences between field types in the spec, and those we support (because of a long-dead argument between people who are no longer involved BTW, but we can't change it now). And it appears to talk about RabbitMQ using signed 64 bit integers only.

*However*, note that it's talking about arbitrary values in AMQP tables (e.g. headers, arguments) not those in method frames.

So yes, this is an ugliness in the spec, but librabbitmq is IMO correct.

**Comment by markwellis** [comment]
@simonmacmullen ok, that clears it up, thanks!

**Comment by noshankus** [comment]
For my money, the spec is incorrect, so it should be a RabbitMQ bug. You cannot have negative delivery tags.

**Comment by noshankus** [comment]
Cool!

So, back to Net-AMQP-RabbitMQ, am I right in saying we need patch using uint64_t for the delivery tag?

**Comment by sbasmanov** [comment]
@noshankus, yes, it means we should use uint64_t. Until @autarch will make patch with Math, You can use my diff.

**Comment by manchicken** [comment]
@sbasmanov, please make a branch and commit it, and pull it into travis-ci
so that we can see if it plays nice with our various supported versions and
such.

**Comment by sbasmanov** [comment]
@manchicken, I'm not very familiar with github, so I forked and created branch. But have no idea how to create pull request to different repo. Here is updated source: https://github.com/sbasmanov/net-amqp-rabbitmq/tree/delivery_tag

**Comment by markwellis** [comment]
@sbasmanov if you create a pull request then travis will test it for you, see https://help.github.com/articles/creating-a-pull-request/

**Comment by sbasmanov** [comment]
@markwellis, I wasn't sure which base to use. There is nothing like 'travis' in pull down list for base fork. That confused me a little. Thanks for helping with this.

**Comment by sbasmanov** [comment]
Updated tests to be consistent with implementation. Travis build 101 passed.

**Comment by manchicken** [comment]
Hey all, I have a branch where I've added this: https://github.com/manchicken/net-amqp-rabbitmq/tree/math-uint64

Please let me know what you think. I have it building in Travis-CI right now, I'll submit a pull request once it passes.

**Comment by autarch** [comment]
So I was looking at the code recently and I realized that it already just uses the `int64_t` and `uint64_t` types in a bunch of places. No one's complained, but I suspect that's simply because they've never hit Perl's limit (which is 2**53 - see http://stackoverflow.com/a/25083776/9832 for a good explanation).

It'd be nice to fix all of these uses at once.

**Comment by manchicken** [comment]
The `Math::Int64` module brings in a typemap which - I believe, haven't looked at it too much - makes an `int64_t` and `uint64_t` instances of `Math::Int64` and `Math::UInt64`.

**Comment by autarch** [comment]
The typemap only applies to values when they cross the boundary between Perl & XS. I'm talking about things like https://github.com/markwellis/net-amqp-rabbitmq/blob/master/RabbitMQ.xs#L608

**Comment by manchicken** [comment]
That's a good point, we could still be truncating there. I didn't think about that. Much thanks.

Any idea what to do with the CApiMaker thing not playing nice with Perl 5.8?

I could just copy the C & H files over into this, but that seems like it's copying code from one repo to bring into another. I could also use a C macro to change the behavior for 5.8 back to what it had been, so that this was only a problem for 5.8 going forward.

**Comment by autarch** [comment]
AFAICT, Module::CAPIMaker only needs 5.10+ because it uses `//=`, but these can all be safely replaced with `||=` because AFAICT `q{}` or `0` are never valid values where it's used.

That said, I think it's fine to copy the .c and .h files into this repo. It simplifies building for people working out of the repo, for one thing.

**Comment by manchicken** [comment]
That makes sense. I'll try that out and maybe that will move this along.

**Comment by manchicken** [comment]
This is closed by pull request #40, yes?

**Comment by markwellis** [comment]
yes sorry! i should pay more attention!
